### ISSUE TITLE
RFC: Ideal gql dx

### ISSUE THREAD

**Issue by Tobbe** [redwoodjs/redwood#1602]
Another thing I want us to collectively think about is what kind of gql DX we want. Without even thinking what exact client we (don't) want, or what's even possible. What would the ideal experience be?

 * What is the minimum feature set we need?
 * What is the cache behavior we want?
 * What's the max bundle size that's acceptable?

We do already allow our users to supply their own Gql client, which is great. But unfortunately it doesn't replace whatever is bundled with RW, it only supplements it. There's a relevant issue that goes in to more details about that here https://github.com/redwoodjs/redwood/issues/1455

**Comment by peterp** [comment]
Hey Tobbe,

I can think of improvements in two places:

1. Discovery of queries.

I think the main point of contention is around mutations an

**Comment by Tobbe** [comment]
> 1. Discovery of queries/ mutations: We facilitate this via the GraphQL vscode extension.

What improvements are needed here?

> 2. An understandable way to update the cache when performing a mutation.

With the different GQL clients you've tried, which one does this the best? How does it work there?

**Comment by peterp** [comment]
> 1. What improvements are needed here?
The extension is great, but it only works when the api side is running via `yarn rw dev`. I think we could improve it by writing the schema to disk and updating the schema to read from this file. Not a train smash, but it might feel a bit more efficient.

I'm looking at things like graphql-zeus, but it really doesn't hit the nail on the head. SDL is pretty great.

In terms of updating the cache, I really hate having to use GraphQL to update the cache. I wish it was a simple flat key-value store, but maybe I'm ignorant to why it isn't.

**Comment by Tobbe** [comment]
> Have you tried react-query?

No, I haven't. I've only used ac, graphql-hooks and my own client.

I'm torn on the caching stuff. The way I'm used to working with remote data is fetching it and putting it in my own (semi-)global state. And from there on out I'm responsible for managing it on my own with optimistic updates, caching/not caching refetches/refreshes etc. Which gives me a lot of control. But on the other hand it's really nice when all the data on a page is automagically updated as soon as I just run a mutation. And once people have seen and used that once, it seems as that's something they expect to just work. Which makes it seem as if we do need a client with some kind of smart caching. Which in my case would leave me with some state in the gql client and some state in redux/context/whatever, which I don't know if I like. I like having just one global data store solution, not two. So that reasoning might lead you to just use the gql client cache for everything instead, which I know Apollo Client supports, and tries to make a case for.

**Comment by jtoar** [comment]
I'll add more thoughts, but wanted to second @peterp's comment real quick

> In terms of updating the cache, I really hate having to use GraphQL to update the cache. I wish it was a simple flat key-value store, but maybe I'm ignorant to why it isn't!

Maybe I'm also ignorant as to why it isn't, but I'm really not sure why it can't just be that way! That's how SWR seems to work:

```js
// tell all SWRs with this key to revalidate
mutate('/api/user')
```

Pretty sure react query makes it simple too. Definitely worth looking more into it.

**Comment by dthyresson** [comment]
Opportunity to rethink GraphQL DX with the Guild and Uri and DP.
